File: tests/api-check.py
"""Run only against a built local Worker with an isolated test database."""
import concurrent.futures
import json
import urllib.error
import urllib.request
import uuid

ORIGIN = 'http://localhost:8788'
credentials = {'email': 'demo@mostrador.test', 'password': 'Mostrador123!'}

def call(path='/api/business', method='GET', payload=None, cookie=None, extra=None):
    headers = {'Content-Type': 'application/json', 'Origin': ORIGIN}
    if cookie:
        headers['Cookie'] = cookie
    headers.update(extra or {})
    request = urllib.request.Request(ORIGIN + path,
        data=json.dumps(payload).encode() if payload is not None else None,
        headers=headers, method=method)
    try:
        response = urllib.request.urlopen(request)
    except urllib.error.HTTPError as error:
        response = error
    with response:
        raw = response.read().decode()
        data = json.loads(raw) if 'application/json' in response.headers.get('Content-Type', '') else raw
        return response.status, data, response.headers, response.geturl()

def login():
    status, _, headers, _ = call('/api/session', 'POST', credentials)
    assert status == 200
    cookie = headers['Set-Cookie']
    assert 'HttpOnly' in cookie and 'SameSite=Lax' in cookie and 'Max-Age=28800' in cookie
    return cookie.split(';', 1)[0]

assert call()[0] == 401
assert call(extra={'oai-authenticated-user-id': 'legacy-owner', 'oai-authenticated-user-email': 'legacy@example.com'})[0] == 401
assert call(cookie='mostrador_demo_session=' + 'f' * 64)[0] == 401
status, root, _, root_url = call('/')
assert status == 200 and root_url == ORIGIN + '/'
assert 'Tu próxima comida' in root and 'Administración' in root and 'Medallón de pollo' in root
assert '/photos/products/medallon-pollo-jamon-queso.jpg' in root
assert call('/catalogo')[3] == ORIGIN + '/'
assert call('/admin')[3] == ORIGIN + '/admin/login'
assert call('/api/session', 'POST', {**credentials, 'password': 'incorrecta'})[0] == 401
assert call('/api/session', 'POST', credentials, extra={'Origin': 'https://other.example'})[0] == 403

cookie = login()
second_cookie = login()
assert call('/admin', cookie=cookie)[3] == ORIGIN + '/admin'
status, business, _, _ = call(cookie=cookie)
assert status == 200 and len(business['products']) == 25

def save(action, state, request_id=None, cookie_value=cookie):
    return call(method='POST', payload={'version': state['version'],
        'requestId': request_id or str(uuid.uuid4()), 'action': action}, cookie=cookie_value)

product = {**business['products'][0], 'price': 987654, 'imageUrl': 'https://example.com/medallon.jpg'}
request_id = str(uuid.uuid4())
status, updated, _, _ = save({'type': 'product', 'product': product}, business, request_id)
assert status == 200, updated
assert save({'type': 'product', 'product': product}, business, request_id)[1]['version'] == updated['version']
assert save({'type': 'product', 'product': product}, business)[0] == 409
assert call(cookie=second_cookie)[1]['products'][0]['price'] == 987654
status, catalog, _, _ = call('/api/business?catalog=1')
assert status == 200 and set(catalog) == {'settings', 'products'}
catalog_product = next(p for p in catalog['products'] if p['id'] == product['id'])
assert catalog_product['price'] == 987654
assert catalog_product['imageUrl'] == product['imageUrl']
assert all(set(p) <= {'id', 'name', 'variety', 'category', 'unit', 'price', 'stock', 'imageUrl'} for p in catalog['products'])

status, hidden, _, _ = save({'type': 'product', 'product': {**product, 'published': False}}, updated)
assert status == 200
assert not any(p['id'] == product['id'] for p in call('/api/business?catalog=1')[1]['products'])
assert save({'type': 'product', 'product': product}, hidden, cookie_value=None)[0] == 401
assert call(method='POST', payload={}, cookie=cookie, extra={'Origin': 'https://other.example'})[0] == 403

payloads = [{'version': hidden['version'], 'requestId': str(uuid.uuid4()), 'action': {
    'type': 'expense', 'name': 'Prueba concurrente', 'category': 'Variable', 'amount': 100, 'paid': False}} for _ in range(2)]
with concurrent.futures.ThreadPoolExecutor(max_workers=2) as pool:
    statuses = list(pool.map(lambda p: call(method='POST', payload=p, cookie=cookie)[0], payloads))
assert sorted(statuses) == [200, 409], statuses

assert call('/api/session', 'DELETE', cookie=cookie)[0] == 200
assert call(cookie=cookie)[0] == 401
assert call('/admin', cookie=cookie)[3] == ORIGIN + '/admin/login'
assert call(cookie=second_cookie)[0] == 200
assert call('/api/session', 'DELETE', cookie=second_cookie)[0] == 200
print('OK: catálogo público renderizado, sesión de prueba, edición compartida, fotos, privacidad, idempotencia, concurrencia, origen y revocación.')

File: tests/api-check.mjs
// Pruebas del contrato HTTP contra el servidor ya construido de la aplicación.
// Levantarlo antes, apuntando a una base descartable:
//   node scripts/build-app.mjs
//   MOSTRADOR_DATABASE_PATH=/tmp/mostrador-qa/business.sqlite PORT=8788 \
//     HOST=127.0.0.1 node dist/standalone/server.js
//   node tests/api-check.mjs
import assert from "node:assert/strict";
import { randomUUID } from "node:crypto";

const ORIGIN = process.env.MOSTRADOR_TEST_ORIGIN || "http://127.0.0.1:8788";
assert.ok(
  ["localhost", "127.0.0.1"].includes(new URL(ORIGIN).hostname),
  "Estas pruebas escriben datos: nunca correrlas contra una base real.",
);

const call = async (path = "/api/business", method = "GET", payload) => {
  const r = await fetch(ORIGIN + path, {
    method,
    headers: payload ? { "Content-Type": "application/json" } : undefined,
    body: payload ? JSON.stringify(payload) : undefined,
    redirect: "manual",
  });
  const text = await r.text();
  let body;
  try { body = JSON.parse(text); } catch { body = text; }
  return { status: r.status, body, headers: r.headers };
};

const save = (action, state, requestId = randomUUID()) =>
  call("/api/business", "POST", { version: state.version, requestId, action });

// La aplicación corre local y sin cuentas: la administración abre en la raíz.
const root = await call("/");
assert.equal(root.status, 200);
assert.ok(typeof root.body === "string" && root.body.includes("<"), "la raíz sirve la aplicación");

// El catálogo público y las sesiones ya no existen.
for (const gone of ["/catalogo", "/admin", "/admin/login", "/api/session"])
  assert.equal((await call(gone)).status, 404, gone + " debería estar dado de baja");
assert.equal((await call("/api/business?catalog=1")).status, 200, "el parámetro catalog ya no cambia la respuesta");

// Los datos se leen sin iniciar sesión, y nunca se filtra el registro interno.
const first = await call();
assert.equal(first.status, 200);
const business = first.body;
assert.ok(business.products.length > 0);
assert.ok(!("completedRequests" in business), "completedRequests no sale al cliente");

// Guardar: la revisión avanza y el mismo requestId no aplica el cambio dos veces.
const product = { ...business.products[0], price: 987654 };
const requestId = randomUUID();
const saved = await save({ type: "product", product }, business, requestId);
assert.equal(saved.status, 200, JSON.stringify(saved.body));
assert.ok(saved.body.version > business.version, "la revisión avanza");
const retry = await save({ type: "product", product }, business, requestId);
assert.equal(retry.body.version, saved.body.version, "reintentar el mismo pedido no duplica");

// Una escritura con la revisión vieja se rechaza en vez de pisar lo guardado.
const stale = await save({ type: "product", product }, business);
assert.equal(stale.status, 409, "revisión vencida rechazada");
assert.equal((await call()).body.products[0].price, 987654);

// Dos escrituras simultáneas: una gana, la otra reintenta.
const current = (await call()).body;
const expense = {
  type: "expense",
  name: "Prueba concurrente",
  category: "Variable",
  amount: 100,
  paid: false,
};
const both = await Promise.all([
  save(expense, current),
  save(expense, current),
]);
assert.deepEqual(
  both.map((r) => r.status).sort(),
  [200, 409],
  "exactamente una de las dos escrituras simultáneas gana",
);

// Límite de tamaño del cuerpo.
const huge = await save(
  { type: "expense", name: "x".repeat(40000), category: "Fijo", amount: 1, paid: false },
  (await call()).body,
);
assert.equal(huge.status, 413, "cuerpo demasiado grande rechazado");

// Datos inválidos vuelven como error de validación, no como caída.
const bad = await call("/api/business", "POST", { version: 0, requestId: "no-es-uuid", action: {} });
assert.equal(bad.status, 400);

console.log("OK: raíz servida, catálogo y sesiones dados de baja, lectura abierta, idempotencia, revisión, concurrencia, tamaño y validación.");
